source/crud/bulk-write.md

```diff
@@ -0,0 +1,724 @@
+classes if their existing result classes deviate from the definitions in this specification (e.g.
+if they contain acknowledgement information, which is not applicable for individual bulk write
+operations). Drivers MAY reuse their existing types for these classes if they match the ones
+defined here exactly.
+
+### Exception
+
+```typescript
+class BulkWriteException {
+    /**
+     * A top-level error that occurred when attempting to communicate with the server or execute
+     * the bulk write. This value may not be populated if the exception was thrown due to errors
+     * occurring on individual writes.
+     */
+    error: Optional<Error>;
+
+    /**
+     * Write concern errors that occurred while executing the bulk write. This list may have
+     * multiple items if more than one server command was required to execute the bulk write.
+     */
+    writeConcernErrors: WriteConcernError[];
+
+    /**
+     * Errors that occurred during the execution of individual write operations. This map will
+     * contain at most one entry if the bulk write was ordered.
+     */
+    writeErrors: Map<Int64, WriteError>;
+
+    /**
+     * The results of any successful operations that were performed before the error was
+     * encountered.
+     */
+    partialResult: Optional<BulkWriteResult>;
+}
+```
+
+## Building a `bulkWrite` Command
+
+The `bulkWrite` server command has the following format:
+
+```json
+{
+    "bulkWrite": 1,
+    "ops": <Array>,
+    "nsInfo": <Array>,
+    "errorsOnly": <Boolean>,
+    "ordered": <Boolean>,
+    "bypassDocumentValidation": <Boolean>,
+    "comment": <Bson>,
+    "let": <Document>,
+    ...additional operation-agnostic fields
+}
+```
+
+Drivers SHOULD use document sequences ([`OP_MSG`](../message/OP_MSG.rst) payload type 1) for the
+`ops` and `nsInfo` fields. Drivers MUST NOT use document sequences when auto-encryption is enabled.
+
+The `bulkWrite` command is executed on the "admin" database.
+
+### Operations
+
+The `ops` field is a list of write operation documents. The first entry in each document has
+the name of the operation (i.e. "insert", "update", or "delete") as its key and the index in the
+`nsInfo` array of the namespace on which the operation should be performed as its value. The
+documents have the following format:
+
+#### Insert
+
+```json
+{
+    "insert": <Int32>,
+    "document": <Document>
+}
+```
+
+Drivers MUST add an `_id` field at the beginning of the insert document if one is not already
+present.
+
+When a user executes a bulk write with an unacknowledged write concern, drivers SHOULD check the
+size of the insert document to verify that it does not exceed `maxBsonObjectSize`. For acknowledged
+bulk writes, drivers MAY rely on the server to return an error if the document exceeds
+`maxBsonObjectSize`. The value for `maxBsonObjectSize` can be retrieved from the selected server's
+`hello` response.
```

Comment:
Suggest adding a check for all documents in `ops` (not just inserts). This may inform a user when there is no chance of a write succeeding. Consider adding a separate section under `### Operations`:

```md
#### Unacknowledged size checks

When a user executes a bulk write with an unacknowledged write concern, drivers MUST check the size of documents in `ops` and error if there is no expected chance of success:

Check that the size of each insert document does not exceed `maxBsonObjectSize`. The value for `maxBsonObjectSize` can be retrieved from the selected server's `hello` response.

Check that the size of other documents does not exceed `maxBsonObjectSize` + 16384 (16 KiB). 16KiB is added to account for command overhead allowed in the server (refer: [SERVER-10643](https://jira.mongodb.org/browse/SERVER-10643)).

For acknowledged bulk writes, drivers MUST NOT do these checks and instead rely on the server for size checks.
```

Notably: C and Python drivers have different existing size checks for OP_MSG document sequences:

[The C driver](https://github.com/mongodb/mongo-c-driver/blob/f6547a375c8d0b9ae7eff898487835ad451ab3d0/src/libmongoc/src/mongoc/mongoc-write-command.c#L643-L647) does not check if the write is unacknowledged, and includes the 16k allowance:

```c
if (slen > max_bson_obj_size + BSON_OBJECT_ALLOWANCE) {
```

[The Python driver](https://github.com/mongodb/mongo-python-driver/blob/b041ca5f7c201f1b0ecb97c517a1ece8d775c00a/pymongo/message.py#L1340) checks if the write is unacknowledged, and does not include the 16k allowance:
```python
unacked_doc_too_large = not ack and (doc_length > max_bson_size)
```

Reply:
Addressed by new `### Size Limits` section.